src/happy.rs

```diff
@@ -0,0 +1,95 @@
+use std::io;
+use std::iter::FusedIterator;
+use std::net::TcpStream;
+use std::net::ToSocketAddrs;
+use std::sync::mpsc::{channel, RecvTimeoutError};
+use std::thread;
+use std::time::{Duration, Instant};
+
+const DEFAULT_CONNECTION_TIMEOUT: Duration = Duration::from_secs(10);
+const RACE_DELAY: Duration = Duration::from_millis(200);
+
+/// This function implements a basic form of the happy eyeballs RFC to quickly connect
+/// to a domain which is available in both IPv4 and IPv6. Connection attempts are raced
+/// against each other and the first to connect successfully wins the race.
+///
+/// If the timeout is not provided, a default timeout of 10 seconds is used.
+pub fn connect<A>(addrs: A, timeout: impl Into<Option<Duration>>) -> io::Result<TcpStream>
+where
+    A: ToSocketAddrs,
+{
+    let timeout = timeout.into().unwrap_or(DEFAULT_CONNECTION_TIMEOUT);
+    let addrs: Vec<_> = addrs.to_socket_addrs()?.collect();
+    let ipv4 = addrs.iter().filter(|a| a.is_ipv4()).cloned();
+    let ipv6 = addrs.iter().filter(|a| a.is_ipv6()).cloned();
+    let order = intertwine(ipv4, ipv6);
+
+    let (tx, rx) = channel();
+    let mut last_err = None;
+
+    let start = Instant::now();
+
+    for addr in order {
+        let tx = tx.clone();
+
+        thread::spawn(move || {
+            debug!("trying to connect to {}", addr);
+
+            let _ = tx.send(TcpStream::connect_timeout(&addr, timeout));
+        });
+
+        match rx.recv_timeout(RACE_DELAY) {
+            Ok(Ok(sock)) => {
+                let end = Instant::now();
+                let span = end - start;
+                debug!("success, took {}ms", span.as_millis());
+
+                return Ok(sock);
+            }
+            Ok(Err(err)) => {
+                debug!("connection error: {} addr={}", err, addr);
+
+                last_err = Some(err);
```

Comment:
I think the first error might be more appropriate if we consider the this the highest priority address? But that is admittedly rather unscientific.

Reply:
But due to the threading, there are no guarantees anyway. So this is really moot. Sorry for the noise.

Reply:
Fixed